File: juno_rs/src/indicators/alma.rs
use super::MA;
use bounded_vec_deque::BoundedVecDeque;
use serde::{Deserialize, Serialize};
use std::cmp::min;

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct AlmaParams {
    pub offset: f64, // 0.85
    pub period: u32,
    pub sigma: Option<u32>, // Calculated from period if None.
}

pub struct Alma {
    pub value: f64,

    weights: Vec<f64>,
    prices: BoundedVecDeque<f64>,

    t: u32,
    t1: u32,
}

impl Alma {
    pub fn new(params: &AlmaParams) -> Self {
        assert!(params.period > 0);
        assert!(0.0 < params.offset && params.offset < 1.0);

        let sigma = match params.sigma {
            Some(sigma) => {
                assert!(sigma > 0);
                sigma
            },
            None => (params.period as f64 / 1.5).floor() as u32,
        };

        let m = (params.offset * (params.period - 1) as f64).floor();
        let s = params.period as f64 * 1.0 / sigma as f64;
        let tmp = (0..params.period)
            .map(|i| (-(i as f64 - m) * (i as f64 - m) / (2.0 * s * s)).exp())
            .collect::<Vec<f64>>();
        let sw: f64 = tmp.iter().sum();
        Self {
            value: 0.0,

            weights: tmp.iter().map(|v| v / sw).collect::<Vec<f64>>(),
            prices: BoundedVecDeque::new(params.period as usize),

            t: 0,
            t1: params.period,
        }
    }
}

impl MA for Alma {
    fn maturity(&self) -> u32 {
        self.t1
    }

    fn mature(&self) -> bool {
        self.t >= self.t1
    }

    fn update(&mut self, price: f64) {
        self.t = min(self.t + 1, self.t1);

        self.prices.push_back(price);

        if self.mature() {
            self.value = self
                .prices
                .iter()
                .zip(self.weights.iter())
                .map(|(p, w)| p * w)
                .sum()
        }
    }

    fn value(&self) -> f64 {
        self.value
    }
}

File: juno/indicators/alma.py
from __future__ import annotations

from collections import deque
from decimal import ROUND_FLOOR, Decimal
from typing import Optional


# Arnaud Legoux Moving Average
class Alma:
    value: Decimal = Decimal('0.0')

    _weights: list[Decimal]
    _prices: deque[Decimal]

    _t: int = 0
    _t1: int

    def __init__(
        self, period: int, sigma: Optional[int] = None, offset: Decimal = Decimal('0.85')
    ) -> None:
        if period < 1:
            raise ValueError(f'Invalid period ({period})')
        if offset < 0 or offset > 1:
            raise ValueError(f'Invalid offset ({offset})')
        if sigma == 0:
            raise ValueError(f'Invalid sigma ({sigma})')

        sig = int(period / Decimal('1.5')) if sigma is None else sigma

        m = (offset * (period - 1)).to_integral_exact(rounding=ROUND_FLOOR)
        s = period * Decimal('1.0') / sig
        tmp = [(-(i - m) * (i - m) / (2 * s * s)).exp() for i in range(period)]
        sw = sum(tmp)
        self._weights = [w / sw for w in tmp]

        self._prices = deque(maxlen=period)

        self._t1 = period

    @property
    def maturity(self) -> int:
        return self._t1

    @property
    def mature(self) -> bool:
        return self._t >= self._t1

    def update(self, price: Decimal) -> Decimal:
        self._t = min(self._t + 1, self._t1)

        self._prices.append(price)

        if self.mature:
            self.value = sum((p * w for p, w in zip(self._prices, self._weights)), Decimal('0.0'))

        return self.value
